### DISCUSSION TITLE
IDetectRobots injection into tag helper constructor

### DISCUSSION THREAD

**Discussion by xperiandri** [OrchardCMS/OrchardCore#8407] [General]
@MatthijsKrempel, @jtkech, @hishamco, @agriffard, do you know why `IDetectRobots` is not just injected into the constructor?

https://github.com/OrchardCMS/OrchardCore/blob/7e8a75360b15dcfab772a7972c90ea37d2d69e33/src/OrchardCore/OrchardCore.ReCaptcha.Core/TagHelpers/ReCaptchaTagHelper.cs#L54

**Comment by hishamco** [comment]
I think it should, I don't see any reason to leave it there

**Reply by hishamco** [reply]
I realize this is needed to get all the registed instances of `IDetectRobots`, but I think we could inject `IEnumerable<IDetectRobots>` instead

**Reply by xperiandri** [reply]
Looks like we can go that way https://medium.com/agilix/asp-net-core-inject-all-instances-of-a-service-interface-64b37b43fdc8
We will try. Thanks

**Comment by MatthijsKrempel** [comment]
Try it and find out, I can’t remember, something vaguely about lifetime registration rings a bell.

**Reply by hishamco** [reply]
May be this line https://github.com/OrchardCMS/OrchardCore/blob/7e8a75360b15dcfab772a7972c90ea37d2d69e33/src/OrchardCore/OrchardCore.ReCaptcha.Core/ActionFilters/Detection/IpAddressRobotDetector.cs#L36

/cc @jtkech @Skrypt

**Reply by Skrypt** [reply]
Probably an exception here where we need absolutely to use the current httpContext.
Though I understand the concern since we should try to avoid injecting the HttpContext as much as we can.
